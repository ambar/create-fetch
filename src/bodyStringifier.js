import querystring from 'querystring'

const isPlainObject = obj =>
  typeof obj === 'object' &&
  obj !== null &&
  Object.getPrototypeOf(obj) === Object.prototype

/**
 * Stringify body
 */
const bodyStringifier = () => fetch => (
  url,
  {body, bodyType, ...options} = {}
) => {
  let itsHeaders = null
  if (isPlainObject(body) || Array.isArray(body)) {
    if (bodyType === 'form') {
      body = querystring.stringify(body)
      itsHeaders = {'content-type': 'application/x-www-form-urlencoded'}
    } else if (bodyType === 'json' || bodyType === undefined) {
      body = JSON.stringify(body)
      itsHeaders = {'content-type': 'application/json'}
    }
  }

  return fetch(url, {
    ...options,
    body,
    headers: {
      ...options.headers,
      ...itsHeaders,
    },
  })
}

export default bodyStringifier
