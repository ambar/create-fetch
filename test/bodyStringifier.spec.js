import bodyStringifier from '../src/bodyStringifier'

const objectSample = {username: 'Jo Jo', lang: '中文', age: undefined}
const arraySample = [1, 2, true]

describe('bodyStringifier', () => {
  const fetch = (...args) => Promise.resolve(args)

  it('should stringify json', async () => {
    const myFetch = bodyStringifier()(fetch)
    expect(await myFetch('/api', {body: objectSample})).toMatchSnapshot()
    expect(await myFetch('/api', {body: arraySample})).toMatchSnapshot()
  })

  it('should stringify form', async () => {
    const myFetch = bodyStringifier()(fetch)
    expect(
      await myFetch('/api', {bodyType: 'form', body: objectSample})
    ).toMatchSnapshot()
    expect(
      await myFetch('/api', {bodyType: 'form', body: arraySample})
    ).toMatchSnapshot()
  })

  it('should not stringify with non-plain-object or non-array', async () => {
    const myFetch = bodyStringifier()(fetch)
    expect(await myFetch('/api', {body: new class Foo {}()})).toMatchSnapshot()
    expect(await myFetch('/api', {body: new FormData()})).toMatchSnapshot()
    expect(await myFetch('/api', {body: new Map()})).toMatchSnapshot()
    expect(await myFetch('/api', {body: true})).toMatchSnapshot()
    expect(await myFetch('/api', {body: null})).toMatchSnapshot()
  })

  it('should not stringify when given a wrong bodyType', async () => {
    const myFetch = bodyStringifier()(fetch)
    expect(
      await myFetch('/api', {bodyType: 'unknown', body: objectSample})
    ).toMatchSnapshot()
  })

  it('should work without options', async () => {
    const myFetch = bodyStringifier()(fetch)
    expect(await myFetch('/api')).toMatchSnapshot()
  })
})
