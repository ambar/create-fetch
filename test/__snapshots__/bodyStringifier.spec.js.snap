// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bodyStringifier should not stringify when given a wrong bodyType 1`] = `
Array [
  "/api",
  Object {
    "body": Object {
      "age": undefined,
      "lang": "中文",
      "username": "Jo Jo",
    },
    "headers": Object {},
  },
]
`;

exports[`bodyStringifier should not stringify with non-plain-object or non-array 1`] = `
Array [
  "/api",
  Object {
    "body": Foo {},
    "headers": Object {},
  },
]
`;

exports[`bodyStringifier should not stringify with non-plain-object or non-array 2`] = `
Array [
  "/api",
  Object {
    "body": Map {},
    "headers": Object {},
  },
]
`;

exports[`bodyStringifier should not stringify with non-plain-object or non-array 3`] = `
Array [
  "/api",
  Object {
    "body": true,
    "headers": Object {},
  },
]
`;

exports[`bodyStringifier should not stringify with non-plain-object or non-array 4`] = `
Array [
  "/api",
  Object {
    "body": null,
    "headers": Object {},
  },
]
`;

exports[`bodyStringifier should stringify form 1`] = `
Array [
  "/api",
  Object {
    "body": "username=Jo%20Jo&lang=%E4%B8%AD%E6%96%87&age=",
    "headers": Object {
      "content-type": "application/x-www-form-urlencoded",
    },
  },
]
`;

exports[`bodyStringifier should stringify form 2`] = `
Array [
  "/api",
  Object {
    "body": "0=1&1=2&2=true",
    "headers": Object {
      "content-type": "application/x-www-form-urlencoded",
    },
  },
]
`;

exports[`bodyStringifier should stringify json 1`] = `
Array [
  "/api",
  Object {
    "body": "{\\"username\\":\\"Jo Jo\\",\\"lang\\":\\"中文\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodyStringifier should stringify json 2`] = `
Array [
  "/api",
  Object {
    "body": "[1,2,true]",
    "headers": Object {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodyStringifier should work without options 1`] = `
Array [
  "/api",
  Object {
    "body": undefined,
    "headers": Object {},
  },
]
`;
